src/main/java/mate/academy/bookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,122 @@
+import mate.academy.bookstore.service.UserService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@RequiredArgsConstructor
+@Service
+public class OrderServiceImpl implements OrderService {
+    private final ShoppingCartService shoppingCartService;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderRepository orderRepository;
+    private final UserService userService;
+    private final OrderItemRepository orderItemRepository;
+
+    @Transactional
+    @Override
+    public OrderDto save(Long userId, CreateOrderRequestDto requestDto) {
+        ShoppingCart cartFromDb = shoppingCartService.getCartByUserId(userId);
+        if (cartFromDb.getCartItems().size() == 0) {
+            throw new IllegalStateException("Cannot proceed with an empty shopping cart.");
+        }
+        Order order = createOrderWithoutItems(userId, requestDto);
+        order = orderRepository.save(order);
+        Set<OrderItem> orderItems = createOrderItemsSet(cartFromDb, order);
+        order.setOrderItems(orderItems);
+        order.setTotal(order.getOrderItems().stream()
+                .map(OrderItem::getPrice)
+                .reduce(BigDecimal.ZERO, BigDecimal::add));
+        shoppingCartService.clearShoppingCart(cartFromDb.getId());
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderDto> findAll(Long userId) {
+        return orderRepository.findAllByUserId(userId).stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public List<OrderItemDto> findAllOrderItemsByOrderId(Long userId, Long orderId) {
+        Order orderFromDb = getOrderFromDbByUserIdAndOrderId(userId, orderId);
+        return orderFromDb.getOrderItems().stream()
+                .map(orderItemMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public OrderItemDto findOrderItemById(Long userId, Long itemId, Long orderId) {
+        Order orderFromDb = getOrderFromDbByUserIdAndOrderId(userId, orderId);
+        return orderFromDb.getOrderItems().stream()
+                .filter(i -> i.getId().equals(itemId))
+                .findFirst()
+                .map(orderItemMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find item by itemId=" + itemId));
+    }
+
+    @Override
+    public OrderDto update(UpdateOrderStatusRequestDto requestDto, Long orderId) {
+        Optional<Order> optionalOrder = orderRepository.findById(orderId);
+        if (optionalOrder.isPresent()) {
+            Order order = optionalOrder.get();
+            order.setStatus(requestDto.status());
+            return orderMapper.toDto(order);
+        } else {
+            throw new EntityNotFoundException("Can't find order by orderId=" + orderId);
+        }
```

Comment:
[throw early](https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement), unnest rest of your code